### PR TITLE
Reference type support in MachInst backend and on AArch64

### PR DESCRIPTION
**Pull request by cfallin**
This PR adds support for reference types in the `MachInst` backend framework and on AArch64. It comes in two parts: the initial support for allowing `R32`/`R64` types to flow through the Wasm program and call boundaries, and the implementation of opcodes that support them (`is_null`, etc.); and then stackmap support to allow GCs to occur while Wasm stackframes are active.

This has been tested with SpiderMonkey and is known to work there with GCs occuring during calls from Wasm into JS. We plan to do further testing in regalloc.rs directly (with our checker) too.

This requires a regalloc.rs version bump (with bytecodealliance/regalloc.rs#79 included) and dep version update; the CI build here will be red until that change is made.

### CODE REVIEW

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs [hunk 1]
```diff
@@ -1085,7 +1085,26 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
...
+            let rd = output_to_reg(ctx, outputs[0]);
+            let rn = input_to_reg(ctx, inputs[0], NarrowValueMode::None);
+            let ty = ctx.input_ty(insn, 0);
+            let (alu_op, const_value) = match op {
+                Opcode::IsNull => {
+                    // cmp rn, #0
+                    (choose_32_64(ty, ALUOp::SubS32, ALUOp::SubS64), 0)
+                }
+                Opcode::IsInvalid => {
+                    // cmn rn, #1
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:1149]: It's actually `cmn`, which is a separate instruction. It compares against the twos-complement value of the RHS. We want `cmp rn, #-1` but `-1` can't be encoded in an `Imm12`, so the idiomatic approach on AArch64 is to do `cmn rn, #-1`.

**Comment by cfallin** [comment]
> This is wrong. `i128` is supposed to be stored as two `i64`, just like `i64` is supposed to be stored as two `i32` on 32bit systems. Storing it in a vector register makes it a lot harder to generate efficient math operations for `i128`.

Fair enough, removed. It was used internally (produced by and matched by the ABI code) to communicate to `store_stack` / `load_stack` that a store/load needs to handle the whole 128-bit vector register, in the case that the regalloc cannot provide a vreg/type associated with a spill.  In other words this match arm would not have been triggered by an `I128` in the incoming CLIF. But the regalloc should not do this in practice (it only lacks a vreg for a spill/reload when it is a reference from a realreg being spilled to a stackmap slot), so we can just remove the match case. I would expect that when we implement true `I128` support, we would lower it to 64-bit ops, and the spill/reload code would not need to reason about `I128`s.

**Comment by cfallin** [comment]
Rebased to latest and updated to use support in `regalloc.rs` that was just merged. The PR to do a new release of regalloc is pending, so this depends on the bytecodealliance/regalloc.rs repo and git hash directly, to let CI work; will update this before a any merge.

@julian-seward1, this should be ready for review now!

**Comment by cfallin** [comment]
(Actually, spoke too quickly re: regalloc -- we just released 0.0.27 so the dep in this PR can use that now.)

**Comment by cfallin** [comment]
Rebased again after changes to `x64` backend.

@julian-seward1 or @bnjbvr, would you mind reviewing this soon before we land additional new-backend PRs, to reduce the rebase overhead somewhat? Thanks!

File: cranelift/codegen/src/isa/aarch64/abi.rs [hunk 2]
```diff
@@ -1070,42 +1129,28 @@ impl ABIBody for AArch64ABIBody {
...
-                    insts.push(sub_inst);
-                }
+                total_sp_adjust += total_stacksize as u64;
             }
         }
 
-        // N.B.: "nominal SP", which we use to refer to stackslots
-        // and spillslots, is *here* (the value of SP at this program point).
+        // N.B.: "nominal SP", which we use to refer to stackslots and
+        // spillslots, is right here.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/abi.rs:1137]: nit: "is right here" is hard to parse.  Does it mean "is correct here"?   Or "at this point in the frame construction, nominal SP = actual SP" etc?  Can you clarify?

File: cranelift/codegen/src/isa/aarch64/abi.rs [hunk 3]
```diff
@@ -1256,15 +1306,30 @@ impl ABIBody for AArch64ABIBody {
...
+        ty: Option<Type>,
+    ) -> Inst {
+        let ty = ty_from_ty_hint_or_reg_class(to_reg.to_reg().to_reg(), ty);
         self.load_spillslot(from_slot, ty, to_reg.map(|r| r.to_reg()))
     }
 }
 
+fn ty_from_ty_hint_or_reg_class(r: Reg, ty: Option<Type>) -> Type {
+    match (ty, r.get_class()) {
+        (Some(t), _) => t,
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/abi.rs:1327]: Can it happen that `Some(t)` is inconsistent with the second (here, ignored) component of the tuple?  It would be good to add a comment explaining the meaning of this function.  In particular why the first component has priority.

File: cranelift/codegen/src/machinst/buffer.rs [hunk 4]
```diff
@@ -1286,6 +1313,18 @@ pub struct MachSrcLoc {
     pub loc: SourceLoc,
 }
 
+/// Record of stackmap metadata: stack offsets containing references.
+#[derive(Clone, Debug)]
+pub struct MachStackMap {
+    /// The code offset at which this stackmap applies.
+    pub offset: CodeOffset,
+    /// The code offset at the *end* of the instruction at which this stackmap
```
Comment (julian-seward1) [inline cranelift/codegen/src/machinst/buffer.rs:1321]: Errr, `*end*` is too vague.  Please specify exactly.  Do you mean: the code offset for the last (highest-addressed) byte of the instruction ..

File: cranelift/codegen/src/machinst/lower.rs [hunk 5]
```diff
@@ -93,6 +94,8 @@ pub trait LowerCtx {
     /// every side-effecting op; the backend should not try to merge across
     /// side-effect colors unless the op being merged is known to be pure.
     fn inst_color(&self, ir_inst: Inst) -> InstColor;
+    /// Determine whether an instruction is a safepoint.
+    fn is_safepoint(&self, ir_inst: Inst) -> bool;
```
Comment (julian-seward1) [inline cranelift/codegen/src/machinst/lower.rs:98]: How is this used?  In particular, will we have to call `is_safepoint` on every insn that is processed?  That would seem to be an undesirable compiler performance hit, if so?  (Tell me it ain't so!)

File: cranelift/codegen/src/machinst/mod.rs [hunk 6]
```diff
@@ -191,6 +191,10 @@ pub trait MachInst: Clone + Debug {
     /// What is the worst-case instruction size emitted by this instruction type?
     fn worst_case_size() -> CodeOffset;
 
+    /// What is the register class used for reference types (GC-observable pointers)? Can
+    /// be dependent on compilation flags.
+    fn ref_type_rc(_flags: &Flags) -> RegClass;
```
Comment (julian-seward1) [inline cranelift/codegen/src/machinst/mod.rs:196]: nit: `..rc` feels a bit too terse here; can you call it `ref_type_regclass` ?

File: cranelift/codegen/src/machinst/vcode.rs [hunk 7]
```diff
@@ -221,13 +255,19 @@ fn is_redundant_move<I: VCodeInst>(insn: &I) -> bool {
     }
 }
 
+/// Is this type a reference type?
+fn is_reftype(ty: Type) -> bool {
```
Comment (julian-seward1) [inline cranelift/codegen/src/machinst/vcode.rs:259]: nano-nit: since most of our users will be coming at this on 64-bit targets, maybe put that first in the disjunction?  That way we can save perhaps 1/3 of a nanosecond per call :-)

File: cranelift/codegen/src/isa/aarch64/abi.rs [hunk 8]
```diff
@@ -400,6 +400,7 @@ fn in_int_reg(ty: ir::Type) -> bool {
     match ty {
         types::I8 | types::I16 | types::I32 | types::I64 => true,
         types::B1 | types::B8 | types::B16 | types::B32 | types::B64 => true,
+        types::R32 | types::R64 => true,
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/abi.rs:407]: Here and below, it concerns me that we are accepting `R32` on a 64-bit target without comment.  IIUC, if anything claiming to be `R32` turns up on a 64-bit target, we have gone way off the rails -- perhaps even to the extent of having a security problem.  In such a circumstance I would much prefer that we bring the system to a halt, even in release builds (== panic in every such `match` on `ty`s); and similarly for x64 stubs, if you have similarly modified them.

Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/abi.rs]: Done, excellent point.

File: cranelift/codegen/src/machinst/lower.rs (hunk 5 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs]: Sorry, this was left over from an earlier iteration that did indeed call `is_safepoint` on every instruction; but we no longer do that, and the function was dead, so I removed it.

File: cranelift/codegen/src/machinst/buffer.rs (hunk 4 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/buffer.rs]: Updated comment, thanks!

File: cranelift/codegen/src/isa/aarch64/abi.rs (hunk 3 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/abi.rs:1339]: The type should always be consistent with the register class; the `Option<Type>` is only missing when the regalloc requests a spill of a realreg (for which it doesn't have a vreg, hence no vcode-level type) at a safepoint, and this will always be an `I64` reg. Added comments to clarify.

**Comment by cfallin** [comment]
Addressed, thanks!
